wallet/core/src/utxo/context.rs

```diff
@@ -352,139 +379,261 @@ impl UtxoContext {
-        Balance::new(mature, pending)
+        let mature: u64 = context.mature.iter().map(|e| e.as_ref().entry.amount).sum();
+        let pending: u64 = context.pending.values().map(|e| e.as_ref().entry.amount).sum();
+
+        // this will aggregate only transactions containing
+        // the final payments (not compound transactions)
+        // and outgoing transactions that have not yet
+        // been accepted
+        let mut outgoing: u64 = 0;
+        let mut consumed: u64 = 0;
+        for tx in context.outgoing.values() {
+            if !tx.is_accepted() {
+                if let Some(payment_value) = tx.payment_value() {
+                    // final tx
+                    outgoing += tx.fees() + payment_value;
+                    consumed += tx.aggregate_input_value();
+                } else {
+                    // compound tx has no payment value
+                    // we skip them, accumulating only fees
+                    // as fees are the only component that will
+                    // reduce the final balance after the
+                    // compound process
+                    outgoing += tx.fees();
+                }
+            }
+        }
+
+        Balance::new(
+            (mature + consumed) - outgoing,
+            pending,
+            outgoing,
+            context.mature.len(),
+            context.pending.len(),
+            context.stasis.len(),
+        )
     }
 
-    pub(crate) async fn handle_utxo_added(&self, utxos: Vec<UtxoEntryReference>) -> Result<()> {
+    pub(crate) async fn handle_utxo_added(&self, utxos: Vec<UtxoEntryReference>, current_daa_score: u64) -> Result<()> {
         // add UTXOs to account set
-        let current_daa_score = self.processor().current_daa_score().expect("daa score expected when invoking handle_utxo_added()");
 
-        for utxo in utxos.iter() {
-            if let Err(err) = self.insert(utxo.clone(), current_daa_score).await {
-                log_error!("{}", err);
+        let mut accepted_outgoing_transactions = AHashSet::new();
+
+        let added = HashMap::group_from(utxos.into_iter().map(|utxo| (utxo.transaction_id(), utxo)));
+        for (txid, utxos) in added.into_iter() {
+            // get outgoing transaction from the processor in case the transaction
+            // originates from a different [`Account`] represented by a different [`UtxoContext`].
+            let outgoing_transaction = self.processor().outgoing().get(&txid);
+
+            let force_maturity_if_outgoing = outgoing_transaction.is_some();
+            let is_coinbase_stasis =
+                utxos.first().map(|utxo| matches!(utxo.maturity(current_daa_score), Maturity::Stasis)).unwrap_or_default();
+
+            for utxo in utxos.iter() {
+                if let Err(err) = self.insert(utxo.clone(), current_daa_score, force_maturity_if_outgoing).await {
+                    log_error!("{}", err);
+                }
```

Comment:
no...    `insert()` actually will never return errors and the only error it can produce is duplicate inserts which are internal and are a valid thing... so this is just a structural leftover where I want to have the convenience of potentially changing the behavior in the future and seeing errors in logs.  it won't reach here.  ...need to think about how to clean up the code up gracefully...  it's basically just an abandoned general error trap left for convenience, but I agree it is confusing...

Reply:
If this error means nothing for the end user,  change the log level, to `warn` maybe

Reply:
These are technically unreachable.  I have added a TODO comment to remove this later and converted an inner (related) error to a warning.  Currently `insert()` will never return an error but I want to keep the error reporting there and the `Result` return type as a placeholder in case the logic needs to change in the near future.  https://github.com/aspectron/rusty-kaspa/commit/03cbccb21c91ad2299a6297397562888f822d99a